"Native" implementation of wxTreeCtrl for wxQT

ffa-grahamdawes opened the pull request:
This pull requests contains an in implementation of wxTreeCtrl.  Previously wxQT was using wxGenericTreeCtrl.

There are a couple of advantages to a native implementation.  Firstly,  wxTreeCtrl is now fully accessibly to screen readers (and test automation software for that matter).  Secondly, the overall appearance of the tree is more consistent with the rest of wxQT.

ffa-grahamdawes committed:
Rename internal tree widget to something more appropriate for
a wxListCtrl.

ffa-grahamdawes committed:
Start "native" implementation of wxTreeCtrl for wxQT

ffa-grahamdawes committed:
Implement basic styles for wxTreeCtrl under wxQT

ffa-grahamdawes committed:
Fix root relationship

ffa-grahamdawes committed:
Add missing d'tor

ffa-grahamdawes committed:
Fix some failing TreeBookCtrl tests cases

ffa-grahamdawes committed:
Add test for wxTreeCtrl::GetItemCount()

ffa-grahamdawes committed:
Add signal handlers to generate wx events for wxTreeCtrl under wxQT

ffa-grahamdawes committed:
Rename wxQtItemEditorFactory to wxQtTreeItemEditorFactory

ffa-grahamdawes committed:
Move wxQtTreeItemEditorFactory to its own header

ffa-grahamdawes committed:
Add EditControl support for wxTreeCtrl under wxQT

ffa-grahamdawes committed:
Better support for images in wxTreeCtrl under wxQT

ffa-grahamdawes committed:
Fix wxTreeCtrl::GetSelections for wxQT

ffa-grahamdawes committed:
Prevent "header" row from showing in wxTreeCtrl under wxQT.

ffa-grahamdawes committed:
Make wxQT wxTreeCtrl selection behave consistently with other wx ports.

ffa-grahamdawes committed:
Fix build warnings in wxTreeCtrl

Commit:
No buttons image list support for Qt as for MSW.

Commit:
Get initial tree control drag and drop working. Create a new type for tree item data, as it needs to be serialisable for drag and drop, and void* is not.

Commit:
Merge branch 'qt_tree_control' of https://github.com/GeoTeric/wxWidgets into qt_tree_control

# Conflicts:
#	src/qt/treectrl.cpp

ffa-grahamdawes committed:
Merge branch 'qt_tree_control' of https://github.com/GeoTeric/wxWidgets into qt_tree_control

Commit:
Clean up the code after merges.

ffa-grahamdawes committed:
Generate event when begin editing tree item.

Commit:
Fix wx drag events to actually be keyed off of the Qt dragging state instead of its "drop" events.
Add "dropped" state to track whether a drop actually occurred or whether the drag ended unsuccessfully.

Commit:
Fix up a comment.

Commit:
Merge branch 'qt_tree_control' of https://github.com/GeoTeric/wxWidgets into qt_tree_control

Commit:
Get rid of logic error - no need for state variables and all that, by just doing things at the right time.
This fixes the problem of the mouse position being wrong during endDrag.

ffa-grahamdawes committed:
Generate wxEVT_TREE_END_LABEL_EDIT when the user has finished editing an item in wxTreeCtrl

ffa-grahamdawes committed:
Store item state in wxTreeCtrl for wxQT

ffa-grahamdawes committed:
Merge branch 'qt_tree_control' of https://github.com/GeoTeric/wxWidgets into qt_tree_control

ffa-grahamdawes committed:
Place holder image is only generated once..

ffa-grahamdawes committed:
Set old tree item for selection change events.

ffa-grahamdawes committed:
Don't add item to tree until its fully constructed (prevents itemChanged signal from firing).

Commit:
Implement "hide root" properly.

Commit:
Fix SetItemBackgroundColour.

Commit:
Clean up code before pull request..

Commit:
Clean up file.

Commit:
Mark a virtual function override as such.

Commit:
Create a simple implementation of GetBoundingRect.

vadz commented on build/files:
```diff
@@ -297,7 +297,16 @@ QT_HDR =
     wx/qt/tooltip.h
     wx/qt/toplevel.h
-    wx/qt/window.h
+		wx/qt/treectrl.h
```
Wrong indentation, please check for hard TABs.

jaynabonne replied:
Fixed.

vadz commented on include/wx/generic/treectlg.h:
```diff
@@ -360,7 +360,7 @@ class WXDLLIMPEXP_CORE wxGenericTreeCtrl : public wxTreeCtrlBase,
 };
 
-#if !defined(__WXMSW__) || defined(__WXUNIVERSAL__)
+#if !defined(__WXMSW__) && ! defined(__WXQT__) || defined(__WXUNIVERSAL__)
```
We should test for `wxHAS_GENERIC_TREECTRL` here.

ffa-grahamdawes replied:
The QT build is similar to the MSW build in that it currently includes both the native and the generic versions of the tree control.  

 Shall I remove the generic version from the QT build?

vadz commented on include/wx/qt/private/treeitemfactory.h:
```diff
@@ -0,0 +1,123 @@
+// Name:        wx/treeitemfactory.h
+// Purpose:     Factory to create text edit controls for the tree items
+// Author:      Graham Dawes
+// Modified by:
```
Very minor, but we don't use this header in the new files, it's a left over from the bad old pre-VCS days.

jaynabonne replied:
Is the preferred header just the same but without the "Modified by" line, or is there something else we should use?

vadz replied:
Sorry for being unclear, the header is fine, it's just the "Modified by:" line that should be omitted.

jaynabonne replied:
Cool. That has been fixed, along with the (one true) date and path issues below.

vadz commented on include/wx/qt/private/treeitemfactory.h:
```diff
@@ -0,0 +1,123 @@
+/////////////////////////////////////////////////////////////////////////////
+// Name:        wx/treeitemfactory.h
```
```suggestion
// Name:        wx/qt/private/treeitemfactory.h
```

vadz commented on include/wx/qt/private/treeitemfactory.h:
```diff
@@ -0,0 +1,123 @@
+// Purpose:     Factory to create text edit controls for the tree items
+// Author:      Graham Dawes
+// Modified by:
+// Created:     02/2019
```
Let's standardize on the One True Date Format:
```suggestion
// Created:     2019-02-07
```

jaynabonne replied:
ISO-8601 all the way!

vadz commented on include/wx/qt/private/treeitemfactory.h:
```diff
@@ -0,0 +1,123 @@
+// Licence:     wxWindows licence
+/////////////////////////////////////////////////////////////////////////////
+
+#ifndef _WX_TREEITEM_FACTORY_H_
```
This convention is not followed everywhere, but normally header guards should include the path:
```suggestion
#ifndef _WX_QT_PRIVATE_TREEITEM_FACTORY_H_
```

vadz commented on include/wx/qt/treectrl.h:
```diff
@@ -21,120 +21,126 @@ class WXDLLIMPEXP_CORE wxTreeCtrl : public wxTreeCtrlBase
+    virtual wxTreeItemId GetFocusedItem() const wxOVERRIDE;
+
+    virtual wxTreeItemId GetItemParent(const wxTreeItemId& item) const wxOVERRIDE;
+     
```
```suggestion

```

vadz commented on samples/treectrl/treetest.cpp:
```diff
@@ -1055,7 +1055,7 @@ void MyTreeCtrl::CreateStateImageList(bool del)
 }
 
-#if USE_GENERIC_TREECTRL || !defined(__WXMSW__)
+#if USE_GENERIC_TREECTRL || (!defined(__WXMSW__) && !defined(__WXQT__))
```
It seems wrong to change it here but not in the other lines testing `USE_GENERIC_TREECTRL`. I'm not sure about the logic here, to be honest, but I think we really need to define `USE_MSW_TREECTRL` once and then just test for it as the intention seems to have some MSW-specific code only.

To test whether this works correctly, you need to change `USE_GENERIC_TREECTRL` to 1 in `samples/treectrl/treetest.h` and build the sample -- does it compile/work?

vadz commented on src/qt/treectrl.cpp:
```diff
@@ -8,11 +8,438 @@
+    {
+    }
+
+    TreeItemDataQt(wxTreeItemData* data) : data(data)
```
```suggestion
    explicit TreeItemDataQt(wxTreeItemData* data) : data(data)
```

vadz commented on src/qt/treectrl.cpp:
```diff
@@ -8,11 +8,438 @@
+
+    wxTreeItemData *data;
+
+    private:
```
It's unusual to have a struct with a private section, I think we should just move `registered` to the ctor in which it is used and then we wouldn't need this private member at all.

vadz commented on src/qt/treectrl.cpp:
```diff
@@ -8,11 +8,438 @@
+    wxTreeItemData *data;
+
+    private:
+    static bool registered;
```
Also, in wx conventions static should be easily greppable for by their special prefix:
```suggestion
    static bool s_registered;
```

vadz commented on src/qt/treectrl.cpp:
```diff
@@ -8,11 +8,438 @@
+bool TreeItemDataQt::registered = false;
+Q_DECLARE_METATYPE(TreeItemDataQt)
+
+QDataStream &operator<<(QDataStream &out, const TreeItemDataQt &WXUNUSED(obj))
```
Why do we implement these operators in such a useless way? Do we have to implement them even if there is no reasonable way to do it? A comment explaining this would be really welcome here.

vadz commented on src/qt/treectrl.cpp:
```diff
@@ -8,11 +8,438 @@
+
+    return totalCount;
+}
+class ImageState
```
Very minor, but please insert a new line to separate the function from the class.

vadz commented on src/qt/treectrl.cpp:
```diff
@@ -8,11 +8,438 @@
+    {
+        for ( int i = wxTreeItemIcon_Normal; i < wxTreeItemIcon_Max; ++i )
+        {
+            m_imageStates[i] = -1;
```
Please prefer using symbolic constants:
```suggestion
            m_imageStates[i] = wxWithImages::NO_IMAGE;
```

vadz commented on src/qt/treectrl.cpp:
```diff
@@ -8,11 +8,438 @@
+        return m_imageStates[index];
+    }
+
+    const int &operator[](size_t index) const
```
Returning int by const reference is weird, why not just
```suggestion
    int operator[](size_t index) const
```

vadz commented on src/qt/treectrl.cpp:
```diff
@@ -8,11 +8,438 @@
+        }
+
+        // QT doesn't update the selection until this signal has been
+        // processed. Deferring this event ensures that
```
Thanks for the explanation, this is very helpful.

Just out of curiosity, is there no possibility to define a "post-signal" handler in Qt, like GTK+ "after" signals, i.e. something that happens after the default handler executes?

vadz commented on src/qt/treectrl.cpp:
```diff
@@ -21,9 +448,9 @@ wxTreeCtrl::wxTreeCtrl(wxWindow *parent, wxWindowID id,
            long style,
            const wxValidator& validator,
-           const wxString& name)
+           const wxString& name) 
```
```suggestion
           const wxString& name)
```

vadz commented on src/qt/treectrl.cpp:
```diff
@@ -56,275 +495,699 @@ void wxTreeCtrl::SetIndent(unsigned int indent)
 {
-    return wxString();
+    if ( !item.IsOk() )
+        return "";
```
```suggestion
        return wxString();
```

vadz commented on src/qt/treectrl.cpp:
```diff
@@ -56,275 +495,699 @@ void wxTreeCtrl::SetIndent(unsigned int indent)
-    return 0;
+    QList<QTreeWidgetItem*> qtSelections = m_qtTreeWidget->selectedItems();
+
+    const size_t numberOfSelections = qtSelections.size();
```
Might want to call `selections.reserve(numberOfSelections)` as a micro optimization.

vadz commented on src/qt/treectrl.cpp:
```diff
@@ -56,275 +495,699 @@ void wxTreeCtrl::SetIndent(unsigned int indent)
+    for ( size_t i = 0; i < numberOfSelections; ++i )
+    {
+        QTreeWidgetItem *item = qtSelections[i];
+        selections.Add(wxQtConvertTreeItem(item));
```
Please prefer std-like methods in the new code:
```suggestion
        selections.push_back(wxQtConvertTreeItem(item));
```

vadz commented on src/qt/treectrl.cpp:
```diff
@@ -56,275 +495,699 @@ void wxTreeCtrl::SetIndent(unsigned int indent)
+
+    do
+    {
+        if (IsVisible(itemid))
```
```suggestion
        if ( IsVisible(itemid) )
```

jaynabonne replied:
Fixed. I even went through all of these last night. I guess I missed a few. :)

vadz commented on src/qt/treectrl.cpp:
```diff
@@ -56,275 +495,699 @@ void wxTreeCtrl::SetIndent(unsigned int indent)
 wxTreeItemId wxTreeCtrl::GetPrevVisible(const wxTreeItemId& item) const
 {
-    return wxTreeItemId();
+    wxCHECK_MSG(item.IsOk(), wxTreeItemId(), wxT("invalid tree item"));
```
Please avoid `wxT()` in the new code. If it's very performance-sensitive you can use `wxS()` but usually it's just fine to not use any macros:
```suggestion
    wxCHECK_MSG(item.IsOk(), wxTreeItemId(), "invalid tree item");
```

jaynabonne replied:
Fixed

vadz commented on src/qt/treectrl.cpp:
```diff
@@ -342,15 +1234,64 @@ wxTreeItemId wxTreeCtrl::DoInsertAfter(const wxTreeItemId& parent,
+    int w, h;
+    GetSize(&w, &h);
+    flags = 0;
+    if ( point.x < 0 ) flags |= wxTREE_HITTEST_TOLEFT;
```
Please don't put if statement body on the same line.

jaynabonne replied:
Fixed.

vadz commented on src/qt/treectrl.cpp:
```diff
@@ -342,15 +1234,64 @@ wxTreeItemId wxTreeCtrl::DoInsertAfter(const wxTreeItemId& parent,
+
+    QTreeWidgetItem *hitItem = m_qtTreeWidget->itemAt(wxQtConvertPoint(point));
+
+    if (hitItem == NULL)
```
```suggestion
    if ( hitItem == NULL )
```

jaynabonne replied:
Fixed (ugh)

vadz commented on src/qt/treectrl.cpp:
```diff
@@ -56,275 +495,699 @@ void wxTreeCtrl::SetIndent(unsigned int indent)
+    {
+        m_qtTreeWidget->removeItemWidget(qTreeItem, 0);
+    }
+
```
It seems strange that `Delete()` doesn't call `DeleteChildren()`. Maybe this happens implicitly somehow, but normally `Delete() == DeleteChildren() + DeleteSelf()`. Does it really work like this here?

ffa-grahamdawes replied:
I'm going to look into this now.   The QT docs are, as always, unclear about whether it handles the delete recursively. 

If it does, do we just need to generate the appropriate events for the children being deleted?

ffa-grahamdawes replied:
Have modified delete to call `DeleteChildren()` for the item being deleted.

vadz replied:
I _think_ we do get events under MSW currently, but ideal would be to add a unit test checking for this.

ffa-grahamdawes replied:
As part of this commit, I modified  the Delete test to include a child item and updated the assertion to check for 2 items being deleted.  

Is this enough or did you have a new test case in mind?

vadz reviewed (commented):
Thanks a lot! Globally this looks excellent to me, there are just a few usual style-related nitpicks, but also a potentially more serious question about `Delete()`. And Travis CI failure seems real enough, could you please look into it?

Thanks again!

Commit:
Fix tab/spacing

Commit:
Fix new header.

Commit:
Fix some stylistic and wxT issues.

ffa-grahamdawes committed:
Fix build errors when building with gcc

ffa-grahamdawes committed:
Delete child items in wxTreeCtrl::Delete under wxQT

ffa-grahamdawes committed:
Fix build warning

ffa-grahamdawes committed:
Fix incorrect icon being drawn in wxTreeCtrl

ffa-grahamdawes committed:
Correct restoration of tree icon logic

ffa-grahamdawes committed:
Fix wxQT wxTreeCtrl drawing issues under Linux

ffa-grahamdawes commented:
I've fixed up a few more issues with this concerning icon drawing once was caused by a merge conflict between Jay and myself the other was caused by the icon drawing not working under Linux.

As far as I can tell I've fixed the CI issues caused by the tree changes and the current failures seem to be issues with the build environment.

vadz commented:
I've restarted one build which seems to have had network problems, but [this problem](https://travis-ci.org/wxWidgets/wxWidgets/jobs/491637034) seems to be real enough -- you're probably using a newer Qt version locally? As usual, we can increase the minimum required version if it's really worth it, but otherwise it would be better to add a version check.

ffa-grahamdawes commented:
I've tested with 5.6 and 5.9 on Windows and Linux.   

 It doesn't seem worth updating the QT version for this change (newer versions require a C++ >= 11 compatible compiler so the whole minimum tool-chain version would need to change) . 

 I'll see if I can find a way to work with older versions of QT 5.

catalinr commented:
> I've tested with 5.6 and 5.9 on Windows and Linux.

apparently `QAbstractItemView::iconSizeChanged()` was added in [5.5.0](https://abi-laboratory.pro/?view=compat_report&l=qt&v1=5.4.2&v2=5.5.0&obj=d4293&kind=abi#Added).

ffa-grahamdawes committed:
Replace use of QTreeWidget::setIcon and QTreeWidget::iconChanged with code that is compatible with our minimum suppported QT version

ffa-grahamdawes committed:
Ensure wxTreeCtrl::DoHitTest updates flag when item is successfully hit.

vadz commented:
@ffa-grahamdawes Do you think this is ready for merging now and/or should it be re-reviewed?

matthew-griffin committed:
Close the Tree Control editor correctly

Ensure that wxTreeItemData is deleted.
Use a QT delegate to create editor and perform custom model update.
Connect to the closeEditor signal to send out end label edit events and decide whether to accept changes.

matthew-griffin commented:
Fixed an issue we had with the text editor never closing, which highlighted the fact that we were also updating the model before sending out events that could veto the change.
Also made sure that the wxTreeItemData is deleted.

matthew-griffin committed:
Fixed build issues

matthew-griffin committed:
Removed spurious indent

matthew-griffin committed:
Support click on selected item to edit

Make QTreeWidgetItems editable and add SelectedClicked as an edit trigger.
Simplify closing the editor as QAbstractItemView already connects to its delegate closeEditor signal.

ffa-grahamdawes committed:
Fix Linux build

matthew-griffin committed:
Fix crash on Linux

matthew-griffin commented:
Since this seems to pass the automated tests now, I think we're happy for this to be considered for merging again.

vadz commented:
@matthew-griffin Thanks for the update!

@catalinr Have you reviewed this already? If you have and don't have any comments, I could avoid (re)reviewing everything here.

catalinr commented on include/wx/qt/private/treeitemfactory.h:
```diff
@@ -0,0 +1,122 @@
+        m_actualParent(actualParent),
+        m_moving(0)
+    {
+        Bind(wxEVT_MOVE, &wxQtListTextCtrl::onMove, this);
```
When is this needed? The first time the control was created (in which case `Unbind()` should probably be called in the handler) or later too (when?)?

I don't know if there are alternatives to this position hack, but it does look a bit suspicious.

catalinr commented on include/wx/qt/private/treeitemfactory.h:
```diff
@@ -0,0 +1,122 @@
+
+private:
+    QWidget* m_actualParent;
+    wxRecursionGuardFlag m_moving;
```
Docs say "wxRecursionGuardFlag object must be declared static or the recursion would never be detected". Are you sure this actually works as intended?

matthew-griffin replied:
The docs appear to use a static flag to ensure it exists for longer than the scope of a single function, with this flag being a member of the class it will live as long as that. Making it static would mean that two instances of the same class could not perform the same action at the same time.

catalinr commented on include/wx/qt/private/treeitemfactory.h:
```diff
@@ -0,0 +1,122 @@
+    virtual QWidget* createEditor(
+        int WXUNUSED(userType),
+        QWidget* parent
+    ) const wxOVERRIDE
```
Is this really too long to be on a single line?

catalinr commented on include/wx/qt/private/treeitemfactory.h:
```diff
@@ -0,0 +1,122 @@
+        QWidget* parent
+    ) const wxOVERRIDE
+    {
+        m_textCtrl = new wxQtListTextCtrl(m_parent, parent);
```
Is there any chance of this being called multiple times and leaking older instances?

catalinr commented on include/wx/qt/private/treeitemfactory.h:
```diff
@@ -0,0 +1,122 @@
+
+private:
+    wxWindow* m_parent;
+    mutable wxTextCtrl* m_textCtrl;
```
Is `mutable` really needed here?

matthew-griffin replied:
I'm not really sure there is an alternative as it needs to be modified in a const function.

catalinr reviewed (changes requested):
First of all this is a really big change and personally I find it difficult to review. IMO some changes should be done first
- it should be broken in a few parts, i.e. "use `QTreeWidget` for `wxTreeCtrl`", and then "use `QTreeWidget` for `wxListCtrl`";
- there should be no unrelated changes, like adding `wxOVERRIDE` everywhere;
- most of your comments have not been addressed;
- I don't know what's the rule for private headers, but should all those implementations be left there?
- I've added a few comments, mostly minor but not all;
- such a PR might even work as intended, but when event their authors find that it's too much to review and change it (otherwise they would've addressed the older comments, right?) then it probably is. And what chance would someone else have later to change that code, even if now it looks like it is better than not having it?

matthew-griffin commented on include/wx/qt/private/treeitemdelegate.h:
```diff
@@ -0,0 +1,73 @@
+        if (!wxTheApp->IsScheduledForDestruction(m_textCtrl))
+        {
+            m_current_model_index = QModelIndex();
+            wxTheApp->ScheduleForDestruction(m_textCtrl);
```
I changed this from a straightforward delete as I found it crashed on Linux, seemed that the focus loss event was still getting through to the deleted control.

matthew-griffin commented on include/wx/qt/private/treeitemfactory.h:
```diff
@@ -0,0 +1,122 @@
+            wxDefaultPosition, wxDefaultSize,
+            wxNO_BORDER),
+        m_actualParent(actualParent),
+        m_moving(0)
```
m_moving? It needs to be initialised as the recursion guard checks whether it is non-zero to determine whether it is in a recursive state.

matthew-griffin committed:
Code review changes

matthew-griffin commented:
I've pushed some changes in response to @catalinr and @vadz earlier comments, I can't speak to all of the earlier work as Graham wasn't able to continue on this feature when it was last looked at. I think the references to lists are because there was some common code that could be shared across list and tree controls. The wxQtListTextCtrl could potentially be renamed to something like wxQtEditTextCtrl and moved to a separate file if that would be better, I don't think the rest of the item factory is now used in this PR.

GeoTericMickW committed:
Stop multiple selection changed messages being sent when calling DeleteChildren on a TreeItem.

vadz commented:
I'm going to merge this because there is a lot of useful work here (thanks!), but there are at least a couple of problems that I see in the treectrl sample with Qt 5.7 under Debian Stretch:

- Immediate cosmetic issue with the icons, which are top-aligned rather than being centered. I'm not sure how this went unnoticed (could this be specific to my Qt version?), but it looks pretty bad.
- Context menu is not shown at all. The event is generated, but the menu seems to be dismissed immediately and never actually appears on screen.
- `IsVisible()` doesn't work correctly, it only checks if the items parent is expanded, but not if it's scrolled off. We need to use `GetBoundingRect()` in it and compare the coordinates with the current window (client) size.

I've fixed another couple of small problems, but I don't know how to fix the ones above, it would be great if you could please look at them.

Additionally, I do share quite a few comments above: e.g. I don't really understand why do we need to use delayed destruction in `destroyEditor()` and what exactly does `wxRecursionGuard` protect against, knowing that it's reset right after we call `move()` anyhow?

ffa-grahamdawes commented:
I'll look at the issues you've reported.

The need for wxRecursionGuard is a bit annoying as QT seems to call closeEditor multiple times. For example programatically closing the editor seems to result it in being called twice. Once for the explicit call and once because the act of closing has caused the editor to lose focus.

The deferred deletion is again, somewhat unfortunate as QT has pending messages for the editor.
